PSReadLine/DynamicHelp.cs

```diff
@@ -0,0 +1,288 @@
+
+            foreach(var token in _singleton._tokens)
+            {
+                if (token.TokenFlags == TokenFlags.CommandName)
+                {
+                    commandName = token.Text;
+                }
+
+                var extent = token.Extent;
+                if (extent.StartOffset <= cursor && extent.EndOffset >= cursor)
+                {
+                    if (token.Kind == TokenKind.Parameter)
+                    {
+                        parameterName = ((ParameterToken)token).ParameterName;
+                        break;
+                    }
+                }
+            }
+
+            _mockableMethods.GetDynamicHelpContent(commandName, parameterName, isFullHelp);
+        }
+
+        private void WriteDynamicHelpBlock(Collection<string> helpBlock)
+        {
+            var bufferWidth = _console.BufferWidth;
+            var colWidth = Math.Min(helpBlock.Max(s => LengthInBufferCells(s)) + 2, bufferWidth);
+            int columns = 1;
+
+            var dynHelp = new MultilineDisplayBlock
+            {
+                Singleton = this,
+                ColumnWidth = colWidth,
+                Columns = columns,
+                Rows = (helpBlock.Count + columns - 1) / columns,
+                HelpItems = helpBlock
+            };
+
+            dynHelp.SaveCursor();
+            dynHelp.DrawHelpBlock(dynHelp);
+
+            Console.ReadKey(intercept: true);
+
+            dynHelp.Clear();
+            dynHelp.RestoreCursor();
+        }
+
+        private void WriteParameterHelp(dynamic helpContent)
+        {
+            Collection<string> helpBlock;
+
+            if (helpContent == null || String.IsNullOrEmpty(helpContent?.Description?[0]?.Text))
+            {
+                helpBlock = new Collection<string>()
+                {
+                    String.Empty,
+                    PSReadLineResources.NeedsUpdateHelp
+                };
+            }
+            else
+            {
+                char c = (char)0x1b;
+
+                string syntax = $"{c}[7m-{helpContent.name} <{helpContent.type.name}>{c}[0m";
+                string desc = "DESC: " + helpContent.Description[0].Text;
+                string details = $"Required: {helpContent.required}, Position: {helpContent.position}, Default Value: {helpContent.defaultValue}, Pipeline Input: {helpContent.pipelineInput}, WildCard: {helpContent.globbing}";
+
+                helpBlock = new Collection<string>
+                {
+                    String.Empty,
+                    syntax,
+                    String.Empty,
+                    desc,
+                    details
+                };
+            }
+
+            WriteDynamicHelpBlock(helpBlock);
+        }
+
+        private class MultilineDisplayBlock : DisplayBlockBase
+        {
+            internal Collection<string> HelpItems;
+
+            public void DrawHelpBlock(MultilineDisplayBlock previewBlock, bool menuSelect = true)
```

Comment:
This method needs some rework, starting from the parameters -- `previousBlock` and `menuSelect` are `Menu` specific, not applicable to `MultilineDisplayBlock`.

Reply:
refactored